server/src/main/java/org/apache/iotdb/db/qp/executor/AbstractQueryProcessExecutor.java

```diff
@@ -63,29 +65,84 @@ protected abstract QueryDataSet processAuthorQuery(AuthorPlan plan, QueryContext
       throws ProcessorException;
 
   private QueryDataSet processDataQuery(QueryPlan queryPlan, QueryContext context)
-      throws StorageEngineException, QueryFilterOptimizationException, PathErrorException, ProcessorException, IOException {
+      throws StorageEngineException, QueryFilterOptimizationException, PathErrorException,
+      ProcessorException, IOException {
+    // deduplicate executed paths
+    List<Path> deduplicatedPaths = new ArrayList<>();
+
     if (queryPlan instanceof GroupByPlan) {
       GroupByPlan groupByPlan = (GroupByPlan) queryPlan;
-      return groupBy(groupByPlan.getPaths(), groupByPlan.getAggregations(),
-          groupByPlan.getExpression(), groupByPlan.getUnit(), groupByPlan.getOrigin(),
-          groupByPlan.getIntervals(), context);
+      List<String> deduplicatedAggregations = new ArrayList<>();
+      deduplicate(groupByPlan.getPaths(), groupByPlan.getAggregations(), deduplicatedPaths,
+          deduplicatedAggregations);
+      return groupBy(deduplicatedPaths, deduplicatedAggregations, groupByPlan.getExpression(),
+          groupByPlan.getUnit(),
+          groupByPlan.getOrigin(), groupByPlan.getIntervals(), context);
     }
 
     if (queryPlan instanceof AggregationPlan) {
-      return aggregate(queryPlan.getPaths(), queryPlan.getAggregations(),
-          queryPlan.getExpression(), context);
+      List<String> deduplicatedAggregations = new ArrayList<>();
+      deduplicate(queryPlan.getPaths(), queryPlan.getAggregations(), deduplicatedPaths,
+          deduplicatedAggregations);
+      return aggregate(deduplicatedPaths, deduplicatedAggregations, queryPlan.getExpression(),
+          context);
     }
 
     if (queryPlan instanceof FillQueryPlan) {
       FillQueryPlan fillQueryPlan = (FillQueryPlan) queryPlan;
-      return fill(queryPlan.getPaths(), fillQueryPlan.getQueryTime(),
-          fillQueryPlan.getFillType(), context);
+      deduplicate(queryPlan.getPaths(), deduplicatedPaths);
+      return fill(deduplicatedPaths, fillQueryPlan.getQueryTime(), fillQueryPlan.getFillType(),
+          context);
     }
-    QueryExpression queryExpression = QueryExpression.create().setSelectSeries(queryPlan.getPaths())
-            .setExpression(queryPlan.getExpression());
+
+    deduplicate(queryPlan.getPaths(), deduplicatedPaths);
+    QueryExpression queryExpression = QueryExpression.create().setSelectSeries(deduplicatedPaths)
+        .setExpression(queryPlan.getExpression());
     return queryRouter.query(queryExpression, context);
   }
 
+  /**
+   * Note that the deduplication strategy must be consistent with that of IoTDBQueryResultSet.
```

Comment:
It is interesting that I did find the modifications of IoTDBQueryResultSet in this PR.

Reply:
Yes, `IoTDBQueryResultSet` is not changed because I fix this bug by modifying the server side.

Reply:
In the original code, the client side `IoTDBQueryResultSet` assumes that the server returns deduplicated query results. However, in the original code, the sever returns query results without the step of deduplication. That's the cause of the bug IOTDB-244.
So, I fix this bug by adding the deduplication step in the server side and uses the same deduplication or map strategy.

Reply:
Got it.
I would appreciate it if you can describe the cause of the bug and your solution briefly at the beginning of the PR. Or you may provide a Jira link and explain it there.